components/brave_service_keys/service_key_utils.cc

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_service_keys/service_key_utils.h"
+
+#include <vector>
+
+#include "base/base64.h"
+#include "base/logging.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_service_keys/buildflags.h"
+#include "crypto/hmac.h"
+#include "crypto/sha2.h"
+
+namespace brave_service_keys {
+
+namespace {
+
+constexpr char kAuthorization[] = "authorization";
+constexpr char kDigest[] = "digest";
+constexpr char kRequestTarget[] = "(request-target)";
+
+}  // namespace
+
+std::pair<std::string, std::string> GetDigestHeader(
+    const std::string& payload) {
+  const std::string value = base::StrCat(
+      {"SHA-256=", base::Base64Encode(crypto::SHA256HashString(payload))});
+  return std::make_pair(kDigest, value);
+}
+
+std::pair<std::string, std::string> CreateSignatureString(
+    base::flat_map<std::string, std::string> headers,
+    const GURL& url,
+    const std::string& method,
+    const std::vector<std::string>& headers_to_sign) {
+  // Create and add the (request-target) header if included
+  if (std::find(headers_to_sign.begin(), headers_to_sign.end(),
+                kRequestTarget) != headers_to_sign.end()) {
+    headers.insert(std::make_pair(
+        kRequestTarget,
+        base::StrCat({base::ToLowerASCII(method), " ", url.PathForRequest()})));
+  }
+
+  std::string header_names;
+  std::string signature_string;
+
+  for (const auto& key : headers_to_sign) {
+    auto header = headers.find(key);
+    if (header != headers.end()) {
+      if (!header_names.empty()) {
+        header_names += " ";
+        signature_string += "\n";
+      }
+      header_names += key;
+      signature_string += key + ": " + header->second;
```

Comment:
`base::StrCat`

Reply:
in particular https://github.com/brave/brave-core/pull/21542/files#r1454216677 applies when you have something like this with `+=` and `+`, but `StrCat` helps there

Reply:
in particular https://github.com/brave/brave-core/pull/21542/files#r1454216677 applies when you have something like this with `+=` and `+`, but `StrCat` helps there